aspnetcore/fundamentals/target-aspnetcore.md

```diff
@@ -0,0 +1,223 @@
+
+Alternatively, .NET Standard 2.0 could be targeted instead of targeting both .NET Core 2.1 and .NET Framework 4.6.1:
+
+[!code-xml[](target-aspnetcore/samples/multi-tfm/alternative-tag-helpers-library.csproj?highlight=4)]
+
+With the preceding project file, .NET Core 2.x and .NET Framework 4.6.1 projects are supported because both target frameworks implement .NET Standard 2.0. Since this library only contains Tag Helpers, it's more efficient to target the specific platforms on which ASP.NET Core runs: .NET Core and .NET Framework. Tag Helpers can't be used by other .NET Standard 2.0-compliant target frameworks such as Unity, UWP, and Xamarin. If your library needs to call platform-specific APIs, target specific .NET implementations instead of .NET Standard. For more information, see [Multi-targeting](/dotnet/standard/library-guidance/cross-platform-targeting#multi-targeting).
+
+## Use an API that hasn't changed
+
+Imagine a scenario in which you're upgrading a middleware library from .NET Core 2.2 to 3.0. The ASP.NET Core middleware APIs being used in the library haven't changed between ASP.NET Core 2.2 and 3.0. To continue supporting the middleware library in .NET Core 3.0, take the following steps:
+
+* Follow the [standard library guidance](/dotnet/standard/library-guidance/).
+* Add a package reference for the API's NuGet package if the assembly doesn't exist in the shared framework.
+
+## Use an API that changed
+
+Imagine a scenario in which you're upgrading a library from .NET Core 2.2 to .NET Core 3.0. An ASP.NET Core API being used in the library has a [breaking change](/dotnet/core/compatibility/breaking-changes) in ASP.NET Core 3.0. Consider whether the library can be rewritten to not use the broken API in all versions.
+
+If you can rewrite the library, do so and continue to target an earlier target framework (for example, .NET Standard 2.0 or .NET Framework 4.6.1) with package references.
+
+If you can't rewrite the library, take the following steps:
+
+* Add a target for .NET Core 3.0.
+* Add a `<FrameworkReference>` element for the shared framework.
+* Use the [#if preprocessor directive](/dotnet/csharp/language-reference/preprocessor-directives/preprocessor-if) with the appropriate target framework symbol to conditionally compile code.
+
+For example, synchronous reads and writes on HTTP request and response streams are disabled by default in ASP.NET Core 3.0. ASP.NET Core 2.2 supports the synchronous behavior by default. Consider a library in which synchronous reads and writes should be enabled in Kestrel. The library should enclose the code to enable synchronous features in the appropriate preprocessor directive. For example:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    // code omitted for brevity
+
+#if NETCOREAPP3_0
+    services.Configure<KestrelServerOptions>(options =>
```

Comment:
Libraries should never do this:
- It's too global
- It's kestrel specific

Instead, they should:
- Avoid sync IO
- Buffer the input or output (need code sample)
- AllowSyncIO where the IO is happening (e.g. https://github.com/aspnet/Benchmarks/blob/21e0050f6c14e0248e02b14395212fce5ae6890b/src/Benchmarks/Middleware/JsonMiddleware.cs#L46-L50)